### PR TITLE
Add negative dimension to transpose

### LINKED ISSUE #657: Transpose using negative dimension
Hi!

It is not a major fix, however I think that x.transpose(-1,-3) should be valid, as it is in numpy:

```
x=np.zeros((3,5,7))
x.transpose(-1,0,-2).shape
```
> (7,3,5)

```
x=torch.FloatTensor(3,5,7)
x.transpose(-1,-2)
```
>Traceback (most recent call last):
>  File "<stdin>", line 1, in <module>
>RuntimeError: out of range at /data/users/soumith/miniconda2/conda-bld/pytorch-
>cuda80-0.1.7_1485449073235/work/torch/lib/TH/generic/THTensor.c:398

Best,
EO

### PR DESCRIPTION
**Pull request by goelhardik**
Fixes #657 
Tested locally.

### CODE REVIEW

**Comment by apaszke** [comment]
I think it'd be better to add annotations to arguments in cwrap, or use the `before_call` attribute, instead of reimplementing the whole function.

**Comment by goelhardik** [comment]
This is the cleanest way I could figure out. Can't think of any other way to modify the `dim`s using just `before_call`. Does this look better?

**Comment by apaszke** [comment]
This looks better. Can you please add a test for it?

**Comment by goelhardik** [comment]
A test asserting that transpose with positive and negative dimensions give the same result is good enough?

**Comment by apaszke** [comment]
Yeah. Just do a for loop from 0 to nDim, and use positive and negative values to ensure that they yield the same result.

**Comment by goelhardik** [comment]
Great. Added a test.

**Review by apaszke — APPROVED** [review]

**Comment by fmassa** [comment]
The builds are failing for CUDA. Maybe you should add `LIBRARY_STATE` in the beginning of the function calls so that it can be CUDA-compatible?

**Comment by goelhardik** [comment]
Adding `LIBRARY_STATE` is not fixing this. For CUDA, it is trying to pass the `LIBRARY_STATE` as the first argument to my function `THTensor_(transpose_neg)`, which does not accept that.

I think either this needs to go to the backend libs as a wrapper like `newTranspose` or maybe a new implementation of the fix using `before_call` etc.

File: torch/csrc/generic/methods/Tensor.cwrap [hunk 1]
```diff
@@ -326,20 +326,36 @@ PyObject * THPTensor_(stride)(PyObject *self, PyObject *args, PyObject *kwargs)
     - THBoolTensor* mask
 ]]
 
+THTensor* THTensor_(transpose_neg)(THTensor *self, THTensor *src, int dim0, int dim1)
```
Comment (colesbury) [inline torch/csrc/generic/methods/Tensor.cwrap]: You'll need to take in a THCState* for CUDA tensors:

```
#if IS_CUDA
THTensor* THTensor_(transpose_neg)(THCState* state, , THTensor *self, THTensor *src, int dim0, int dim1)
#else
THTensor* THTensor_(transpose_neg)(THTensor *self, THTensor *src, int dim0, int dim1)
#endif
```

Then, in the calls to THTensor_(...)(...) you should pass `LIBRARY_STATE` before self:

```
THTensor_(transpose)(LIBRARY_STATE self, src, dim0, dim1);
```

**Review by colesbury — COMMENTED** [review]
Are you sure you can't do this with `before_call`? That seems like it would be much simpler.

File: torch/csrc/generic/methods/Tensor.cwrap (hunk 1 above)
Comment (goelhardik) [inline torch/csrc/generic/methods/Tensor.cwrap]: Great! This should make everything work. Let me make the changes. Thanks!

**Comment by soumith** [comment]
Thanks a lot Hardik!

**Pull request merged**